rfcs/ils-0104-library-kpis.md

```diff
@@ -0,0 +1,424 @@
+### Periodic Stats
+A considered alternative for creating periodic stats would be to create an entire new class of stats in `invenio-stats`.
+It involves creating a new indexer class `PeriodicEventsIndexer` that does not consume events from a queue, but rather generates events on a schedule.
+This was scrapped, as it would introduce unnecessary complexity to `invenio-stats`.
+
+### KPIs
+
+#### KPI 2 and 3.2 - Alternative approach to measure waiting time and loan duration
+An alternative approach to measure the waiting time and loan duration would be do count the loans that are respectively waiting or active at the end of each day.
+This would give a more direct measure of how many loans are active or waiting over time.
+However, this does not allow to compute an average, as it is unsure what divisor to use.
+
+#### KPI 2 and 3.2 - Directly computing the average
+For KPI 2 and 3.2, we could also directly compute the average in the stats aggregation, instead of returning the sum and count and letting the dashboard compute the average.
+However, this allows for less flexibility in what date granularity the average is computed on (e.g. daily, weekly, monthly).
+
+## Unresolved questions
+
+### Periodic Stats
+
+#### Should the periodic stats safeguard against duplicate entries?
+Currently, the periodic stats do not safeguard against multiple events being generated for the same time period.
+E.g. if the `process_periodic_stats` method is ran twice in a single day, it will generate two events for the same day.
+This can lead to unexpected results, as the aggregations will then aggregate over both events.
+A possible solution could be the aggregation of the periodic events always taking the average.
+This way, multiple events for the same time period would not change the result of the aggregation.
+
+
+### KPIs
+
+#### KPI 3.1 - Extracting for loan creation method
+It is still not clear what is the best way to listen for loan creations.
+A challenge we face is that we have to extract the loan creation method.
+The different endpoints called to create a loan are:
+  - interlibrary loans `/borrowing-requests/<pid>/patron-loan/create`
+  - self checkout `/loans/self-checkout`
+  - manual creation `/circulation/loans/request`
+
+We could call a new signal `loan_created` from those endpoints with a parameter indicating the creation method.
+
+Alternatively, we could just listen to the signal `after_record_insert` from `invenio_records`, filter for loans and only during event generation or preprocessing extract the creation method. (Unsure if possible)
+
+#### Median vs. Average
+Should the aggregation of the loan duration and waiting time also be done on a median basis?
+ISO 11620:2023 recommends the median as a more robust measure for the waiting time.
+An additional aggregation could be added.
+
+#### Aggregation period
+We aggregate most stats on a daily basis.
+An exception of this are the loan durations and waiting times, which are aggregated monthly.
```

Comment:
what is the motivation behind montly aggregation? what do we gain/lose by aggregating daily too?